### DISCUSSION TITLE
NEF for classification: choosing a different cost function?

### DISCUSSION THREAD

**Discussion by cnrgbot** [nengo/nengo-zoo#282] [General Discussion]
> _Archived from the Nengo Discourse forum._ Originally posted by **Seanny123** on 2016-10-08. (Forum topic #144, post #1 / id 327)

When you're trying to train a neural network for classification, you usually use [Cross-Entropy Error instead of Mean Squared Error](https://jamesmccaffrey.wordpress.com/2013/11/05/why-you-should-use-cross-entropy-error-instead-of-classification-error-or-mean-squared-error-for-neural-network-classifier-training/) as a measure of the error between the desired and actual network output. When training with the NEF, we always use the Mean Squared Error.

1. Has anyone tried to you use the Cross-Entropy Error instead? If not,
2. How would you go about implementing it? Would you need to make your own solver or would there be some other way?

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **tbekolay** on 2016-10-18. (Forum topic #144, post #2 / id 351)

I'm not aware of anyone using cross-entropy error, and I'm also not aware of anyone who's done NEF-style optimization for direct classification. It's very likely possible, but yeah, why classify when you can regress?

One way to do something similar with Nengo now would be to basically output 0 or 1 for classified / not classified as a dimension for each possible class. That's what I did in [this paper](http://mindmodeling.org/cogsci2013/papers/0058/paper0058.pdf) to get the classification accuracy metrics in Table 3.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Seanny123** on 2016-10-18. (Forum topic #144, post #3 / id 352)

> why classify when you can regress

I don't think I understand the difference between these mathematically, which might be the core of my problem.

> One way to do something similar with Nengo now would be to basically output 0 or 1 for classified / not classified as a dimension for each possible class.

Yeah, I've been using one-hot encoding vectors for my classification experiments.

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **user_752** on 2016-10-20. (Forum topic #144, post #4 / id 358)

Eric has been doing exactly these things for MNIST for fun :)  I'd recommend talking to him.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Seanny123** on 2017-06-02. (Forum topic #144, post #5 / id 882)

FYI, Eric's code for classifying MNIST can be found in [the nengo_examples repository](https://github.com/nengo/nengo_extras/tree/master/examples/vision).
